src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -1040,3 +1147,67 @@ fsm_promote_standby(Keeper *keeper)
 
 	return true;
 }
+
+
+/*
+ * fsm_promote_standby is used in several situations in the FSM transitions and
+ * the following actions are needed to promote a standby:
+ *
+ *    start_postgres
+ * && promote_standby
+ * && add_standby_to_hba
+ * && create_replication_slot
+ * && disable_synchronous_replication
+ * && keeper_drop_replication_slots_for_removed_nodes
+ *
+ * So we reuse fsm_disable_replication() here, rather than copy/pasting the same
+ * bits code in the fsm_promote_standby() function body. If the definition of
+ * the fsm_promote_standby transition ever came to diverge from whatever
+ * fsm_disable_replication() is doing, we'd have to copy/paste and maintain
+ * separate code path.
```

Comment:
Well we still reuse `fsm_disable_replication()`, it's just not the only thing we do... do you have a rewrite in mind?

Reply:
I reworked the comment, we are actually copy/pasting from another transition function and now make that clear. I still believe that we should copy/paste rather than invent an intermediate function and call it from two places with an argument for the expected state of the primary, but maybe we could already do that.